회원가입, 로그인, 내계정 페이지 기능 추가

altmit opened the pull request:
## Description
- 회원가입, 로그인, 내계정 페이지 기능 추가
- 로컬스토리지에 유저 정보 유무에 따라서 자동 로그인 상태 관리 추가
- zustand로 유저 정보 관리하는 state 추가
## Key changes
![1](https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/41321198/5b8c41fc-12a5-47f7-b8c9-7d69a5f25062)
![2](https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/41321198/9abd0fb8-10c5-411f-b3fe-6bc76779b311)
![3](https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/41321198/ea637a01-1841-4aef-88f5-70510961ea37)

- github 로그인 버튼 추가했습니다.
- 회원가입, 내 계정 페이지에서 사용하는 ProfileButton 컴포넌트 추가했습니다.
- Button 컴포넌트 만들 당시에 disabled를 생각하지 못해 disabled 옵션을 줘도 blur 처리가 적용되지 않습니다.
  - 그래도 disabled 상태에서는 버튼 비활성화는 됩니다.
- profileButton 우측 상단에 x 버튼을 누르면 alert이 나오며 확인을 누르면 이미지를 삭제합니다.
- 회원가입 성공시 자동으로 로그인 페이지로 전환됩니다.
- 내 계정 페이지에서 로그아웃 누르면 자동으로 로그인 페이지로 전환됩니다.

## To reviewers
- 내 계정 페이지에서 이미지 수정 시 ProfileButton 아래 저장, 취소 버튼이 있으면 좋을 것 같아 추후에 구현해 보겠습니다.
- x버튼이 검정색이라 조금 칙칙해 보이네요.
- github 로그인 버튼을 검정 배경에 흰 글씨와 지금 적용된 색상중 고민 했는데 어떤게 좋을까요

## Issue
- #22

altmit committed:
#feat #22 : 로그인/회원가입 페이지 기능 추가

- msw login, signup 추가
- 로그인 회원가입 정규식 추가로 submit 버튼 disabled 추가

altmit committed:
Merge branch 'dev' into feat/#22-authPage

altmit committed:
Merge branch 'dev' into feat/#22-authPage

altmit committed:
feat #22 : 회원가입, 내 계정 페이지에서 사용하는 이미지 업로드 버튼 구현

altmit committed:
feat #22 : 로그인 유무를 판단하는 useEffect 추가

altmit committed:
feat #22 : 로컬스토리지 관리하는 파일 추가

altmit committed:
feat #22 : 유저 관련 정보를 관리하는 store 추가

altmit committed:
feat #22 : 회원가입, 로그인, 내 계정 페이지 기능 추가

silvertae commented on src/App.tsx:
```diff
@@ -1,12 +1,28 @@
 export function App() {
+  const { setAccessToken, setProfileImageUrl, setUserName } = useAuthStore();
+
+  useEffect(() => {
```
아예 앱 실행시 localStorage에 저장된 토큰과 유저 정보를 set 해주는거 좋네요. 이렇게 해야지만 새로고침때 로그인이 안풀리고 유지돼서 좋더라구요!

silvertae commented on src/components/ProfileButton.tsx:
```diff
@@ -0,0 +1,99 @@
+  const [isOpenAlert, setIsOpenAlert] = useState(false);
+
+  const onOpenAlert = (event: React.MouseEvent) => {
+    event.stopPropagation();
```
`event.stopPropagation()`를 적용하지 않았을 때 어떤 오류가 발생해서 추가하신건가요?

altmit replied:
ProfileButton 자체를 클릭하면 이미지를 추가하는 uploader가 나옵니다.
그리고 ProfileButton 내부에는 이미지 삭제 버튼도 존재합니다.

이때 이미지 삭제 버튼을 클릭하면 ProfileButton 클릭 까지 이벤트 전파가 생겨서 alert과 이미지 uploader가 함께 나와서 이벤트 전파를 막았습니다.

altmit committed:
refactor #22 : octocat svg 추가

altmit commented:
Oauth login button에 octocat svg 추가
![1](https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/41321198/fbc21978-7050-437f-968f-83864f16c86a)

silvertae commented on src/components/ProfileButton.tsx:
```diff
@@ -0,0 +1,99 @@
+import { Icon } from './icon/Icon';
+
+type ProfileButtonProps = {
+  file: File | undefined;
```
props로 file 자체를 받아오는 건가요? 컴포넌트 내부에서는 `(file || backgroundImage)`로 프로필 사진 존재 여부를 판단할때만 사용되는 것 같은데, 그러면 file 타입으로 전체를 넘겨주지 않고 boolean 등으로 존재 여부만 알아도 되지 않을까 해서요.

altmit replied:
넵 그렇게도 사용 가능할 것 같네요.
file 자체가 type이 File | undefined라서 추가 작업을 안하고 file 자체로 유무를 판단 했습니다.

silvertae commented on src/mocks/authHandlers.ts:
```diff
@@ -0,0 +1,20 @@
+
+const user = {
+  username: 'testUser',
+  profileImageUrl: 'https://avatars.githubusercontent.com/u/41321198?v=4',
```
무슨 사진이지 해서 들어가봤는데 칼 든 병아리였네요. 🔪🐥

silvertae commented on src/mocks/handlers.ts:
```diff
@@ -1,7 +1,9 @@
   rest.get('/api/test', (_, res, ctx) => {
     return res(ctx.status(200), ctx.json('hello'));
   }),
+  ...authHandlers,
```
오 이런식으로 용도에 따른 핸들러로 분리해서 둘 수 도 있군요. 지금까지 MSW 계속 쓰면서 이런 식으로 사용할 생각을 못했네요.

altmit replied:
한곳에서 모두 선언하면 아무리 mock data라고 해도 관리가 너무 힘들더라고요

silvertae commented on src/stores/authStore.ts:
```diff
@@ -0,0 +1,26 @@
+  userName: '',
+  profileImageUrl: '',
+  setAccessToken: (token: string) => set(() => ({ accessToken: token })),
+  setUserName: (name: string) => set(() => ({ userName: name })),
+  setProfileImageUrl: (url: string) => set(() => ({ accessToken: url })),
```
아래 localStorage에 set 함수랑 이름이 똑같아서 순간 헷갈렸네요. 
이건 zustand로 만든 store에 저장해두는 거군요.
localStorage와 다르게 userInfo 객체가 아니라 userName, profileImageUrl로 나누신건 중첩된 객체에서 상태변경을 하려면 복잡해져서 따로 두신건가요?

altmit replied:
저도 이부분 사용하는 곳에서 이름이 똑같아서 헷갈린적이 있습니다.
수정이 필요할 것 같은데 이름을 어떻게 지어야 할 지 모르겠네요 😅

altmit replied:
그리고 userName, profileImageUrl을 따로 놓은 이유는 변경에 있어서 귀찮을 것 같아서 구분해서 사용 했습니다.

silvertae replied:
`getUserInfoFromLocalStorage()` 이렇게하면 너무 길어지긴 하네요 🥲

silvertae commented on src/page/auth/SignUpPanel.tsx:
```diff
@@ -34,28 +45,98 @@ export function SignUpPanel({ closePanel }: SignUpPanelProps) {
+
+      if (file && file.type.startsWith('image/')) {
+        setFile(file);
+        const reader = new FileReader();
```
FileReader 라는 Web API도 있나보네요. 저도 Jay 코드 보고 찾아봤는데 이번에 공부 좀 해보겠습니다. 🤓

altmit replied:
저도 이미지 업로드를 이번에 처음 해봐서 구글링과 gpt를 함께 사용하며 처음 알았습니다

silvertae commented on src/page/auth/LoginPage.tsx:
```diff
@@ -0,0 +1,180 @@
+  const onChangePassword = (event: ChangeEvent<HTMLInputElement>) => {
+    setPassword(event.target.value);
+  };
+
+  const submit = async () => {
+    const res = await fetch('/api/login', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({ id, password }),
+    });
+
+    if (res.ok) {
+      const data = await res.json();
+      const { accessToken, ...userInfo } = data;
+
+      setAccessToken(accessToken);
+      setUserInfo(userInfo);
+
+      authStore.setAccessToken(accessToken);
+      authStore.setUserName(userInfo.userName);
+      authStore.setProfileImageUrl(userInfo.profileImageUrl);
+    }
+  };
```
여기선 아직 react-query를 사용하진 않으셨나보네요. 저도 주말에 공부하려고 했는데 못했던... 추후에 적용해봐도 될 것 같네요.
그리고 리뷰어님 피드백 반영해서 요청 관련 코드가 많아지기 전에 api endpoint도 상수로 미리 만들어두고 사용하면 좋을것 같아요.
api 폴더에 하나 만들까요?

altmit replied:
우선은 fetch로 간단한 api호출을 적용한 상태입니다.

api 관련 pr을 따로 작성할 예정 이라서 그 부분에서 한번에 수정하려합니다.

silvertae commented on src/page/auth/MyProfilePage.tsx:
```diff
@@ -0,0 +1,120 @@
+      } else {
+        event.target.value = '';
+
+        // TODO : toast alert 추가?
```
toast alert도 시간되면 공통컴포넌트로 한 번 만들어보면 좋겠네요. 그런데 toast 알람은 시간되면 사라져야 하나요? 그냥 모달이나 알럿과 차이점을 한 번 찾아봐야겠네요.

altmit replied:
toast alert을 mobile UI에서 에러나 가입 성공 같은 간단한 alert을 잠깐 몇 초 보여주고 사라지는 alert으로 많이 사용하 더라고요.

그래서 추가하면 좋을 것 같다고 생각했습니다.

silvertae reviewed (approved):
zustand 스토어 활용도 잘 하신것 같고, ProfileButton 따로 만들어 두신 것도 잘 봤습니다. 공부가 많이 됐네요. 고생많으셨습니다!

altmit committed:
refactor #22 : authStore 변경

- set 함수명에 setState를 접두사로 붙임
-name, imageUrl 추가 set 함수를 setStateUserInfo 하나로 변경
